editor/src/clj/editor/workspace.clj

```diff
@@ -246,29 +256,66 @@ ordinary paths."
 (defn snapshot-cache [workspace]
   (g/node-value workspace :snapshot-cache))
 
-(defn- is-plugin-file? [resource]
+(defn- is-plugin-clojure-file? [resource]
   (contains? #{"clj"} (resource/ext resource)))
 
-(def dbg-workspace (atom -1))
-
-(defn- find-editor-plugins [workspace]
-  (let [resources (filter is-plugin-file? (g/node-value workspace :resource-list))]
+(defn- find-clojure-plugins [workspace]
+  (let [resources (filter is-plugin-clojure-file? (g/node-value workspace :resource-list))]
     resources))
 
 (defn- load-plugin! [workspace resource]
   ; TODO Handle Exceptions!
   (log/info :msg (str "Loading plugin" (resource/path resource)))
-  (let [plugin (load-string (slurp resource))]
-    (plugin workspace))
+  (let [plugin-fn (load-string (slurp resource))]
+    (plugin-fn workspace))
   (log/info :msg (str "Loaded plugin" (resource/path resource))))
 
 (defn load-editor-plugins! [workspace added]
-  (reset! dbg-workspace workspace)
   (let [added-resources (set (map resource/proj-path added))
-        plugin-resources (find-editor-plugins workspace)
+        plugin-resources (find-clojure-plugins workspace)
         plugin-resources (filter (fn [x] (contains? added-resources (resource/proj-path x))) plugin-resources)]
     (dorun (map (fn [x] (load-plugin! workspace x)) plugin-resources))))
 
+; Determine if the extension has plugins, if so, it needs to be extracted
+
+(defn- is-jar-file? [resource]
+  (contains? #{"jar"} (resource/ext resource)))
+
+(defn- is-plugin-file? [resource]
+  (string/includes? (resource/proj-path resource) "/plugins/"))
+
+(defn- find-plugins-shared-libraries [workspace]
+  (let [resources (filter is-plugin-file? (g/node-value workspace :resource-list))]
+    resources))
+
+(defn unpack-resource! [workspace resource]
+  (let [target-path (plugin-path workspace (resource/proj-path resource))
+        parent-dir (.getParentFile ^File target-path)
+        input-stream (clojure.java.io/input-stream resource)]
+    (when-not (.exists parent-dir)
+      (.mkdirs parent-dir))
+    (io/copy input-stream target-path)))
+
+; It's important to use the same class loader, so that the type signatures match
+(def class-loader (clojure.lang.DynamicClassLoader. (.getContextClassLoader (Thread/currentThread))))
+
+(defn load-class! [class-name]
+  (Class/forName class-name true class-loader))
```

Comment:
We currently have no way to know which workspace we are using.
Not sure we'll ever need it but it's good to know about the problem, since potentially you can have two workspaces which use different versions of the same extension (classes) and a race condition arises.

Reply:
For the time being, it is safe to assume there will be only one workspace.